### PR TITLE
feat: support specifying a custom driver location

### PR DESCRIPTION
**Pull request by mxschmitt**
Relates #59

### CODE REVIEW

**Comment by mxschmitt** [comment]
@StarpTech let me know what you think about the API design. I tried to be backwards compatible, at least with the Run() method.

**Comment by StarpTech** [comment]
This change will definitely help me to install the driver at a specific place once but it would be great if we install the driver at a fixed place by default e.g user-dir like https://playwright.dev/docs/installation#managing-browser-binaries

**Comment by mxschmitt** [comment]
> https://playwright.dev/docs/installation#managing-browser-binaries

Yeah good point. A key requirement there would be as you said, that the version gets used as a cache key. The downside of that is, why I didn't implement it in the first place that the cleanup is way harder to implement. So we can't really say which project is not used anymore.

**Comment by StarpTech** [comment]
> Yeah good point. A key requirement there would be as you said, that the version gets used as a cache key. The downside of that is, why I didn't implement it in the first place that the cleanup is way harder to implement. So we can't really say which project is not used anymore.

But that the same way playwright goes right? It shouldn't be changed so frequently and they are still free to manage a custom place. Additionally, `~/.cache` is subject to a different lifetime.

**Comment by mxschmitt** [comment]
> > Yeah good point. A key requirement there would be as you said, that the version gets used as a cache key. The downside of that is, why I didn't implement it in the first place that the cleanup is way harder to implement. So we can't really say which project is not used anymore.
> 
> But that the same way playwright goes right? It shouldn't be changed so frequently and they are still free to manage a custom place. Additionally, `~/.cache` is subject to a different lifetime it's an established cache directory.

I agree, it's better than it's currently and in the future some cleanup can be included. Will update the PR so its using the same base directory as Playwright is using.

**Comment by mxschmitt** [comment]
@StarpTech I updated the implementation, would be great if you could go over it and let me know what you think, thanks and Merry Christmas!

**Comment by StarpTech** [comment]
@mxschmitt I will do and Merry Christmas too!

File: run.go [hunk 1]
```diff
@@ -14,36 +15,59 @@ import (
...
 }
 
+func getDefaultCacheDirectory() (string, error) {
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", fmt.Errorf("could not get user home directory: %v", err)
+	}
+	switch runtime.GOOS {
+	case "windows":
+		return filepath.Join(userHomeDir, "AppData", "Local", "ms-playwright"), nil
```
Comment (StarpTech) [inline run.go:52]: I'm not sure if we should store files in the same folder as playwright. It's cleaner to maintain a separate folder `playwright-go` on the root of the cache directory.

**Review by StarpTech — CHANGES_REQUESTED** [review]
LGTM

**Comment by StarpTech** [comment]
This might be interesting regarding cache invalidation https://github.com/microsoft/playwright/blob/3eef2548e4d55db5970a81582bc0276d814c42da/src/install/installer.ts#L90-L99 As far I understood playwright doesn't allow you to use different versions on the same host. Unused browsers for the version are cleaned up.

File: run.go (hunk 1 above)
Comment (mxschmitt) [reply run.go:52]: I agree, lets maybe call it ms-playwright-go then so we don't "mess" up the original directory.

**Comment by mxschmitt** [comment]
> This might be interesting regarding cache invalidation https://github.com/microsoft/playwright/blob/3eef2548e4d55db5970a81582bc0276d814c42da/src/install/installer.ts#L90-L99 As far I understood playwright doesn't allow you to use different versions on the same host. Unused browsers for the active version are cleaned up.

Yup we could follow up with something like that in the future. But since it doesn't change that often as the browsers I think its fine if we do it in a separate PR.

File: run.go (hunk 1 above)
Comment (StarpTech) [reply run.go:52]: I think `ms` stands for Microsoft :smile: I don't mind :+1:

Comment (mxschmitt) [reply run.go]: I think its fine because then the directories are directly next to each other.

**Comment by StarpTech** [comment]
After this PR the cache location is `~/.cache/ms-playwright-go/playwright-go/0.170.0-next.1608058598043`  I think we agreed on `~/.cache/ms-playwright-go/0.170.0-next.1608058598043` ?

**Comment by mxschmitt** [comment]
> After this PR the cache location is `~/.cache/ms-playwright-go/playwright-go/0.170.0-next.1608058598043` I think we agreed on `~/.cache/ms-playwright-go/0.170.0-next.1608058598043` ?

Yup also noticed, fixed this behaviour in #66.

**Pull request merged**